### ISSUE TITLE
Windows: hid_enumerate() skips devices that have already been opened

### ISSUE THREAD

**Issue by mrpippy** [signal11/hidapi#23]
I am trying to use the latest hidapi with the Qt-based HID Bootloader software written by Microchip (which is included in the latest Microchip Application Library and located at Microchip Solutions v2011-07-14/USB/Device - Bootloaders/HID/Software - Cross Platform/)

This app runs hid_enumerate on a timer (once a second by default), always searching for its given VID and PID. If the device is found, it opens it with hid_open() and starts communicating. However, the next time hid_enumerate() runs, it is unable to open the device (because of the '0' passed for the share value), thus unable to read the VID/PID, and returns a linked list without the device included. 
To the app, this means that the device is now disconnected, and it calls hid_close(). Then the device is detected when hid_enumerate() runs, and the loop starts over.

I was able to fix this problem by removing the first (0 share value) CreateFileA call in open_device (so CreateFileA is always called with FILE_SHARE_READ|FILE_SHARE_WRITE). 
Is there any better way to solve this problem though? Another way to detect device disconnection maybe, or (even better) a more lightweight or callback-based way to get device connection/disconnection events?

**Comment by TheOnlyJoey** [comment]
So, this is still a issue with the current Github version or 0.8rc.
We are testing with HMD hid devices running hidtext.exe and when the device is opened by another process it still does not show up (Tested on Windows 7 SP1 and Windows 10 with latest versions).
This is odd since setupapi shouldn't require device access and devcon works reporting these devices as well.

@signal11 Any chance enumeration could be looked at again? as far as i understand opening the devices should not be necessary when using the enumeration features in setupapi on Windows.

**Comment by signal11** [comment]
Feel free to pick around at the code and figure out what's going on. Some devices behave strangely. Maybe the other process is opening it in exclusive mode.

**Comment by TheOnlyJoey** [comment]
Regardless of exclusive mode or not, enumeration should always be possible (enumeration functions of setupapi can always enumerate every device). Maybe rewrite hidapi to use the setupapi features directly?

**Comment by signal11** [comment]
HIDAPI does use setupapi features directly. Have a look at the code for yourself. If you know setupapi, the code isn't hard to follow. Step through it with a debugger and see if you can isolate where the issue is. There's not a lot I can do on my end because I don't have your device, and different devices do some different (and weird) things.

**Comment by TheOnlyJoey** [comment]
So, i spend 2 days debugging this, and it seems HIDAPI is doing some extra unnecessary steps after enumerating the devices.
This includes already opening handles and such which is one of the primary reasons why certain exclusive opened devices can not be enumerated.

I created a test application to show the enumeration functionality, i will also work on a PR with a new enumeration function using this method.
http://pastebin.com/xPspjM8Q if there is any feedback please let me know.

**Comment by aspalmer** [comment]
When I try to open devices exclusively it doesn't work until I disable/re-enable the device in the device manager. It then works for open open. Is that likely to be fixed by your change? When the problem is occurring I can see an extra symlink to the USB device in Process Explorer.

**Comment by signal11** [comment]
What are these "unnecessary steps" you speak of?

**Comment by aspalmer** [comment]
@signal11 I think his point is that you can get all the info you need in hid_enumerate with SetupDi calls without calling open_device or any of the functions that use write_handle.

**Comment by signal11** [comment]
Which versions of windows will that work back to?
